staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator_test.go

```diff
@@ -0,0 +1,367 @@
+	var delegationHandlers []http.Handler
+	delegate1 := &openAPIHandler{openapi: &spec.Swagger{
+		SwaggerProps: spec.SwaggerProps{
+			Paths: &spec.Paths{
+				Paths: map[string]spec.PathItem{
+					"/apis/foo/v1": {},
+				},
+			},
+		},
+	}}
+	delegationHandlers = append(delegationHandlers, delegate1)
+
+	s := buildSpecAggregator(delegationHandlers, mux)
+
+	apiService := &v1.APIService{
+		Spec: v1.APIServiceSpec{
+			Service: &v1.ServiceReference{Name: "dummy"},
+		},
+	}
+	apiService.Name = "apiservice"
+
+	handler := &openAPIHandler{
+		returnErr: true,
+		openapi: &spec.Swagger{
+			SwaggerProps: spec.SwaggerProps{
+				Paths: &spec.Paths{
+					Paths: map[string]spec.PathItem{
+						"/apis/apiservicegroup/v1": {},
+					},
+				},
+			},
+		},
+	}
+
+	s.AddAPIService(apiService, handler)
+
+	swagger, err := fetchOpenAPI(mux)
+	if err != nil {
+		t.Error(err)
+	}
+	expectPath(t, swagger, "/apis/foo/v1")
+	expectNoPath(t, swagger, "/apis/apiservicegroup/v1")
+
+	handler.returnErr = false
+	err = s.UpdateAPIService(apiService.Name)
+	if err != nil {
+		t.Error(err)
+	}
+	swagger, err = fetchOpenAPI(mux)
+	if err != nil {
+		t.Error(err)
+	}
+	expectPath(t, swagger, "/apis/foo/v1")
+	expectPath(t, swagger, "/apis/apiservicegroup/v1")
+}
+
+type openAPIHandler struct {
+	openapi   *spec.Swagger
+	returnErr bool
+}
+
+func (o *openAPIHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if o.returnErr {
+		w.WriteHeader(500)
+		return
+	}
+	data, err := json.Marshal(o.openapi)
+	if err != nil {
+		panic(err)
+	}
+	http.ServeContent(w, r, "/openapi/v2", time.Now(), bytes.NewReader(data))
+	return
+}
+
+func fetchOpenAPI(mux *http.ServeMux) (*spec.Swagger, error) {
+	server := httptest.NewServer(mux)
+	defer server.Close()
+	client := server.Client()
+
+	req, err := http.NewRequest("GET", server.URL+"/openapi/v2", nil)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	body, err := io.ReadAll(resp.Body)
+
+	swagger := &spec.Swagger{}
+	if err := swagger.UnmarshalJSON(body); err != nil {
+		return nil, err
+	}
+	return swagger, err
+}
+
+func buildSpecAggregator(delegationHandlers []http.Handler, mux common.PathHandler) *specAggregator {
+	downloader := NewDownloader()
+	aggregatorSpec := &spec.Swagger{
+		SwaggerProps: spec.SwaggerProps{
+			Paths: &spec.Paths{
+				Paths: map[string]spec.PathItem{
+					"/apis/apiregistration.k8s.io/v1": {},
+				},
+			},
+		},
+	}
+	s := buildSpecAggregatorForLocalServices(&downloader, aggregatorSpec, delegationHandlers, mux)
+	return s
+}
+
+func expectPath(t *testing.T, swagger *spec.Swagger, path string) {
+	if _, ok := swagger.Paths.Paths[path]; !ok {
+		t.Errorf("Expected path %s to exist in aggregated paths", path)
+	}
+}
+
+func expectNoPath(t *testing.T, swagger *spec.Swagger, path string) {
+	if _, ok := swagger.Paths.Paths[path]; ok {
+		t.Errorf("Expected path %s to be omitted in aggregated paths", path)
+	}
+}
```

Comment:
Tests look good in general, I think we should be able to have a good way to benchmark that package now.

Reply:
We can isolate benchmarking with crd controller, aggregator, etc in unit tests. Some changes are interlocked (eg: making CRDs lazily aggregated will somewhat affect how the aggregator aggregates everything) and will require a benchmark with more components or pprof.